lxd/instance/drivers/driver_qemu_bus.go

```diff
@@ -0,0 +1,142 @@
+package drivers
+
+import (
+	"fmt"
+	"strings"
+)
+
+type qemuBusEntry struct {
+	bridgeDev int // Device number on the root bridge.
+	bridgeFn  int // Function number on the root bridge.
+
+	dev string // Existing device name.
+	fn  int    // Function number on the existing device.
+}
+
+type qemuBus struct {
+	name string           // Bus type.
+	sb   *strings.Builder // String builder to use.
+
+	portNum int // Next available port/chassis on the bridge.
+	devNum  int // Next available device number on the bridge.
+
+	rootPort *qemuBusEntry // Current root port.
+
+	entries map[string]*qemuBusEntry // Map of qemuBusEntry for a particular shared device.
+}
+
+func (a *qemuBus) allocateRoot() *qemuBusEntry {
+	if a.rootPort == nil {
+		a.rootPort = &qemuBusEntry{
+			bridgeDev: a.devNum,
+		}
+		a.devNum++
+	} else {
+		if a.rootPort.bridgeFn == 7 {
+			a.rootPort.bridgeFn = 0
+			a.rootPort.bridgeDev = a.devNum
+			a.devNum++
+		} else {
+			a.rootPort.bridgeFn++
+		}
+	}
+
+	return a.rootPort
+}
+
+// allocate() does any needed port allocation and  returns the bus name,
+// address and whether the device needs to be configured as multi-function.
+//
+// The group parameter allows for grouping devices together as a single
```

Comment:
@stgraber This part "The group parameter allows for grouping devices together as a single multi-function device. It automatically keeps track of the number of functions already used and will allocate a new device as needed." seems to conflict, as it initially suggests providing a non-empty `group` argument will cause it to allocate ports using a single multifunction device, but then suggests that additional devices could be allocated.

Reply:
Perhaps rewording to "The group parameter allows for grouping devices together on one or more multi-function ports. It automatically keeps track of the number of functions already used and will allocate new multi-function ports as needed"